### PR TITLE
sql: improve tenant records

### PR DESCRIPTION
**Pull request by knz**
supersedes #95574, #95581, #95655
Epic: CRDB-21836

RFC for context: https://github.com/cockroachdb/cockroach/pull/96147

**TLDR** This commit contains the following changes:
- rename "state" to "data_state", "active" to "ready"
- stored, non-virtual columns for "name", "data_state", "service_mode"
- deprecate the column "active" since it mirrors "data_state'
- move `descpb.TenantInfo` to new package `mtinfopb`.
- new statements `ALTER TENANT ... START SERVICE EXTERNAL/SHARED`,
  `STOP SERVICE` to change the service mode.

Details follow.

**rename TenantInfo.State to DataState, "ACTIVE" to "READY"**

We've discovered that we'd like to separate the readiness of the data
from the activation of the service. To emphasize this, this commit
renames the field "State" to "DataState".

Additionally, the state "ACTIVE" was confusing as it suggests that
something is running, whereas it merely marks the tenant data as ready
for use. So this commit also renames that state accordingly.

**new tenant info field ServiceMode**

Summary of changes:
- the new TenantInfo.ServiceMode field indicates how to run servers.
- new syntax: `ALTER TENANT ... START SERVICE EXTERNAL/SHARED`,
  `ALTER TENANT ... STOP SERVICE`.
- tenants created via `create_tenant(<id>)`
  (via CC serverless control plane) start in service mode EXTERNAL.
- other tenants start in service mode NONE.
- need ALTER TENANT STOP SERVICE before dropping a tenant.
  - except in the case of `crdb_internal.destroy_tenant`
    for compat with CC serverless control plane.

**make the output columns of SHOW TENANT lowercase**

All the SHOW statements report status-like data in lowercase.
SHOW TENANT(s) should not be different.

**use actual SQL columns for the TenantInfo fields**

Release note: None

### CODE REVIEW

**Comment by knz** [comment]
I amended the commit to move `descpb.TenantInfo` to its own package as suggested by Andrew.

File: pkg/multitenant/mtinfopb/info.proto
Comment (dt) [inline pkg/multitenant/mtinfopb/info.proto:38]: shouldn't this have a `reserved 3;` replacement?

File: pkg/multitenant/mtinfopb/info.proto
Comment (dt) [inline pkg/multitenant/mtinfopb/info.proto:68]: There is some precedent for "Next ID: x" with various capitalization as the last line of a message elsewhere in the codebase.

File: pkg/multitenant/mtinfopb/info.proto
Comment (dt) [inline pkg/multitenant/mtinfopb/info.proto:67]: mentioned above but I think we typically don't reuse an ID and instead `reserved x;` on the line where it was to make the removal/addition pairing obvious in the diff.

File: pkg/multitenant/mtinfopb/info.proto
Comment (dt) [inline pkg/multitenant/mtinfopb/info.proto:95]: I'm not sure I immediately see what we get by nesting these specific cols in this `ExtraColumns` message vs just having name, id, etc be fields of `TenantInfoWithUsage` (and renaming it to TenantInfoExtended)?

**Review by knz — COMMENTED** [review]
> Reviewed 51 of 51 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @herkolategan, @michae2, @smg260, and @stevendanna)

___
*[`pkg/multitenant/mtinfopb/info.proto` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-67:-NMVkBf89t1dDU--c6Kl:b-ihll8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L67)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

mentioned above but I think we typically don't reuse an ID and instead `reserved x;` on the line where it was to make the removal/addition pairing obvious in the diff.
</blockquote></details>

This one can be reused because it was added in a 23.1 PR which we are (mindfully) choosing to ignore the existence of.
___
*[`pkg/multitenant/mtinfopb/info.proto` line 68 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-68:-NMVkHJF93oji5HPD_ai:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L68)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

There is some precedent for "Next ID: x" (with various capitalization) as the last line of a message elsewhere in the codebase.
</blockquote></details>

Done.
___
*[`pkg/multitenant/mtinfopb/info.proto` line 95 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-95:-NMVkS-kDj_pE0ryCU9u:bf9brn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L95)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'm not sure I immediately see what we get by nesting these specific cols in this `ExtraColumns` message vs just having name, id, etc be fields of `TenantInfoWithUsage` (and renaming it to TenantInfoExtended)?
</blockquote></details>

Because ExtendedTenantInfo contains only a subset of TenantInfoWithUsage.

If we had just 1 struct with everything, it wouldn't be clear in the various tenant mutation functions what to do with all the other usage-related fields.
___
*[`pkg/sql/catalog/descpb/tenant.proto` line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVkfLMEIzI2Ns-NbAa:-NMVkfLMEIzI2Ns-NbAb:b-as0359) ([raw file](https://github.com/cockroachdb/cockroach/blob/35b2a26c52b710954945acf173d6d6a83088b404/pkg/sql/catalog/descpb/tenant.proto#L59)):*
> shouldn't this have a reserved 3; replacement?

No - we are (mindfully) choosing to ignore the PR that introduced it.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @herkolategan, @knz, @michae2, @smg260, and @stevendanna)

___
*[`pkg/multitenant/mtinfopb/info.proto` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-67:-NMVlAWABkUrjVBk8rOJ:b-wb79x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L67)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This one can be reused because it was added in a 23.1 PR which we are (mindfully) choosing to ignore the existence of.
</blockquote></details>

thanks, that makes since but I think it is potentially more confusing to future readers than just reserving 3 as it had been used

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @herkolategan, @michae2, @smg260, and @stevendanna)

___
*[`pkg/multitenant/mtinfopb/info.proto` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-67:-NMVmsfe2KaC-HAZHXai:b3718tn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L67)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

thanks, that makes since but I think it is potentially more confusing to future readers than just reserving 3 as it had been used
</blockquote></details>

Since all the fields after 3 have been added by other PRs that have not yet been released, maybe I could just renumber all of them?

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @herkolategan, @knz, @michae2, @smg260, and @stevendanna)

___
*[`pkg/multitenant/mtinfopb/info.proto` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-67:-NMVsCqV2IvAtljqi1nE:b-er7zas) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L67)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Since all the fields after 3 have been added by other PRs that have not yet been released, maybe I could just renumber all of them?
</blockquote></details>

technically I guess that is true in this specific but I don't really see why not just reserve 3 like we would in the general case but whatever

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @herkolategan, @michae2, @smg260, and @stevendanna)

___
*[`pkg/multitenant/mtinfopb/info.proto` line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA3-r7-67:-NMVvo2G2SF8pxVQdxO-:bgrb9sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c1c5f0b7b13f8eaf7c670a17ca6e5ff99a464a9/pkg/multitenant/mtinfopb/info.proto#L67)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

technically I guess that is true in this specific but I don't really see why not just reserve 3 like we would in the general case but whatever 
</blockquote></details>

Varint encoding of proto tags is more efficient when the values are small -- in particular we get very efficient coding for all values below 15. "Giving up" tag 3 which has not been used makes me sad. I wouldn't care as much if it was 113.

**Comment by knz** [comment]
@dt @stevendanna welcoming your expertise on this failed roach
```
/artifacts/acceptance/version-upgrade/run_1
(mixedversion.go:298).Run: mixed-version test failure while running step 13 (run "run backup"): pq: backupccl.retrieveAllTenantsMetadata: column "system.public.tenants.name" does not exist
test random seed:                       -433825789743187912
binary versions:                        [1: 22.2, 2: 1000022.2-30, 3: 22.2, 4: 22.2]
cluster versions before failure:        [1: 22.2, 2: 22.2, 3: 22.2, 4: 22.2]
cluster versions after failure:         [1: 22.2, 2: 22.2, 3: 22.2, 4: 22.2]
```

It's telling me that `retrieveAllTenantsMetadata` is looking at the system.tenants schema pre-upgrade. Why is that? What should I fix?

File: pkg/upgrade/upgrades/tenant_table_migration_test.go
Comment (ajwerner) [inline pkg/upgrade/upgrades/tenant_table_migration_test.go:61]: should this test maybe use the tenants table after the upgrade? Perhaps create a tenant and then query the table and validate the structure?

File: pkg/sql/tenant_update.go
Comment (ajwerner) [inline pkg/sql/tenant_update.go:75]: this does mostly works, but technically is not bulletproof. The better thing to do would be to resolve the table descriptor and see if it has the new columns in a writable state. I won't block this PR over that though. To do that, consider making the `isql.Txn` into a `descs.Txn` in order to access the descriptors.

File: pkg/upgrade/upgrades/tenant_table_migration_test.go
Comment (ajwerner) [inline pkg/upgrade/upgrades/tenant_table_migration_test.go:71]: If you take another iteration of the code can you put a TODO here that says we should be bootstrapping into the old version and avoid this dance altogether?

File: pkg/upgrade/upgrades/tenant_table_migration.go
Comment (ajwerner) [inline pkg/upgrade/upgrades/tenant_table_migration.go:80]: The assumption baked into this migration is that no other tenants have names. Is that legit? Maybe comment that somewhere

File: pkg/upgrade/upgrades/tenant_table_migration.go
Comment (ajwerner) [inline pkg/upgrade/upgrades/tenant_table_migration.go:85]: My sense is you could just as well use a testing knob and have the base case use the const. This is okay though.

**Review by ajwerner — COMMENTED** [review]
Nothing major from me. Thanks for moving the protobuf file.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @herkolategan, @michae2, @smg260, and @stevendanna)

___
*[`pkg/sql/tenant_update.go` line 75 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMUkLQzEDzvhxAMWfaa-r13-75:-NMZWlmw7HBjF3lzsEnC:b-vc2hi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/91ccefec8f5802e31253b41eddee07cdfa8188e9/pkg/sql/tenant_update.go#L75)):*
> this does mostly works, but technically is not bulletproof.

1) I don't fully understand why -- if my reading is correct, the version can only be bumped after the migration is complete, at which point the descs are updated. Where is this understanding incorrect?
2) we have used this pattern in numerous places before -- use a cluster version gate to choose how to access a system table. If what you say is true, this means we have a wider problem on our hands than just this PR. I'm not sure this PR is the place to fix it.
___
*[`pkg/upgrade/upgrades/tenant_table_migration.go` line 80 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMUkLQzEDzvhxAMWfae-r13-80:-NMZX4auD9vyRQsi2u8U:b9qxkf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/91ccefec8f5802e31253b41eddee07cdfa8188e9/pkg/upgrade/upgrades/tenant_table_migration.go#L80)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The assumption baked into this migration is that no other tenants have names. Is that legit? Maybe comment that somewhere
</blockquote></details>

Yes it is assumed because names can only be introduced after the version field has been updated. Adding comment.
___
*[`pkg/upgrade/upgrades/tenant_table_migration_test.go` line 61 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVTi8cnal-J-sHauCi-r13-61:-NMZYWnH6l17HToyl82R:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/91ccefec8f5802e31253b41eddee07cdfa8188e9/pkg/upgrade/upgrades/tenant_table_migration_test.go#L61)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should this test maybe use the tenants table after the upgrade? Perhaps create a tenant and then query the table and validate the structure?
</blockquote></details>

Good idea. Done.
___
*[`pkg/upgrade/upgrades/tenant_table_migration_test.go` line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVTi8cnal-J-sHauCi-r13-71:-NMZY4Fz0yfDh2wrAdW2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91ccefec8f5802e31253b41eddee07cdfa8188e9/pkg/upgrade/upgrades/tenant_table_migration_test.go#L71)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

If you take another iteration of the code can you put a TODO here that says we should be bootstrapping into the old version and avoid this dance altogether?
</blockquote></details>

Done.

File: pkg/multitenant/mtinfopb/info.go
Comment (stevendanna) [inline pkg/multitenant/mtinfopb/info.go:66]: Do we want the zero-value to be DataStateReady?

File: pkg/sql/logictest/testdata/logic_test/tenant
Comment (stevendanna) [inline pkg/sql/logictest/testdata/logic_test/tenant:261]: Can we add some tests that modify the tenant service mode?

**Review by stevendanna — COMMENTED** [review]
Overall this looks pretty reasonable to me, but I'd like to see a test or two to encourage us to write test when we make future modifications to this.

  I'm not in love with having some fields in the protobuf and some as full columns. But, I don't have a good suggestion here so I defer to the outcome of the previous discussions.

On the topic of `reserved 3` in the protobuf. It seems like if we are going to assume that this never shipped by _not_ reserving it, then there isn't much reason to not just do the renumbering now. That is, either we are going to follow the backwards-compatible change rules or not. If we aren't, then let's not leave a gap in the IDs for later.

**Review by knz — COMMENTED** [review]
>  I'd like to see a test or two to encourage us to write test when we make future modifications to this.

Done.

>I'm not in love with having some fields in the protobuf and some as full columns. But, I don't have a good suggestion here so I defer to the outcome of the previous discussions.

We've discussed elsewhere and I believe the consensus is to push us to go more towards separate columns over time.

> On the topic of reserved 3 in the protobuf. It seems like if we are going to assume that this never shipped by not reserving it, then there isn't much reason to not just do the renumbering now.

:+1: Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-:-NMekI7b3mE7cn-558QK:b-kt9o94)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @herkolategan, @michae2, @smg260, and @stevendanna)

___
*[`pkg/sql/logictest/testdata/logic_test/tenant` line 261 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMUkLQzEDzvhxAMWfaJ-r14-261:-NMelFWD8lB_YkFbT_Bf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ffc6a827ac170ff70f41b93ef5634dcad71fd19/pkg/sql/logictest/testdata/logic_test/tenant#L261)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Can we add some tests that modify the tenant service mode?
</blockquote></details>

Done.
___
*[`pkg/multitenant/mtinfopb/info.go` line 66 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMVd56DhclQfKq6CxA2-r14-66:-NMekTF_0DP7CwoDO9QD:bcpri9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ffc6a827ac170ff70f41b93ef5634dcad71fd19/pkg/multitenant/mtinfopb/info.go#L66)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we want the zero-value to be DataStateReady?
</blockquote></details>

I don't know - what do you think? I'm putting it on "Add" now but I think it doesn't matter. We don't use the zero value elsewhere IIRC.

File: pkg/sql/parser/sql.y
Comment (stevendanna) [inline pkg/sql/parser/sql.y:16103]: I've recently learned that we should potentially also add these new keywords to the bare_keywords list.  But, I plan to do a pass for all the keywords added in 23.1 soon.

File: pkg/ccl/backupccl/testdata/backup-restore/restore-tenants
Comment (stevendanna) [inline pkg/ccl/backupccl/testdata/backup-restore/restore-tenants:18]: We could update this test file to show that servce_state is properly restored during a tenant restore.

**Review by stevendanna — COMMENTED** [review]
Thanks for adding the tests. I left a few more comments, but I trust you to decide which direction to take them.

I see you already had some discussion with @dt around the shape of the protobufs.  It feels a little odd that the main type that most Go code will want to traffic in is TenantInfoExtended.  Perhaps just changing the name of that struct to something like TenantRecord or something might make it feel a bit more reasonable.

I think it is likely that this isn't the final form of how we manage/represent the service state of a tenant, but I don't think this will be too hard to change in the future if necessary.

**Review by stevendanna — APPROVED** [review]

**Review by knz — COMMENTED** [review]
> Perhaps just changing the name of that struct to something like TenantRecord or something might make it feel a bit more reasonable.

Good call. I renamed as follows:

- TenantInfo -> ProtoInfo
- TenantInfoWithUsage_ExtraColumns -> SQLInfo
- ExtendedTenantInfo -> TenantInfo

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-:-NMf28QS9FnIgFo8CrF_:bkh1z0b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @herkolategan, @michae2, @smg260, and @stevendanna)

___
*[`pkg/sql/parser/sql.y` line 16103 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMUkLQzEDzvhxAMWfaN-r15-16103:-NMewkHvCtFXrnG2MEG_:b-m9nqxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/61f2cd976832133ab8ee6c5bb5072da785b1c8d4/pkg/sql/parser/sql.y#L16103)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I've recently learned that we should potentially also add these new keywords to the bare_keywords list.  But, I plan to do a pass for all the keywords added in 23.1 soon.
</blockquote></details>

Thanks I'll let you do this then.
___
*[`pkg/ccl/backupccl/testdata/backup-restore/restore-tenants` line 18 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/95691#-NMUkLQzEDzvhxAMWf_z-r15-18:-NMewosp1eUHU4kTmgcF:br9m6mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/61f2cd976832133ab8ee6c5bb5072da785b1c8d4/pkg/ccl/backupccl/testdata/backup-restore/restore-tenants#L18)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We could update this test file to show that servce_state is properly restored during a tenant restore.
</blockquote></details>

Let's do it! Done.

**Comment by knz** [comment]
This PR will conflict with #95013.
This latter PR is scheduled to merge tonight. I will give it a chance, then rebase mine tomorrow morning before merging.

**Comment by knz** [comment]
bors r=stevendanna,dt,ajwerner

**Comment by knz** [comment]
bors r=stevendanna,dt,ajwerner

**Pull request merged**
